shuup/product_discounts/models.py

```diff
@@ -0,0 +1,164 @@
+            (Q(start_datetime__isnull=True) | Q(start_datetime__lte=current_local_dt)) &
+            (Q(end_datetime__isnull=True) | Q(end_datetime__gte=current_local_dt)) &
+            ~Q(id__in=excluded_discount_ids_based_on_exceptions) &
+            (
+                Q(valid_hours__isnull=True) |
+                Q(
+                    valid_hours__weekday=current_local_weekday,
+                    valid_hours__from_hour__lte=current_local_time,
+                    valid_hours__to_hour__gt=current_local_time,
+                )
+            )
+        )
+
+        if shop:
+            query &= Q(shops=shop)
+
+        return self.filter(query)
+
+
+@python_2_unicode_compatible
+class ProductDiscount(models.Model, MoneyPropped):
+    name = models.CharField(
+        null=True, blank=True, max_length=120, verbose_name=_("name"),
+        help_text=_("The name for this discount. Used internally with discount lists for filtering."))
+    identifier = InternalIdentifierField(unique=True)
+    created_by = models.ForeignKey(
+        settings.AUTH_USER_MODEL, blank=True, null=True,
+        related_name="+", on_delete=models.SET_NULL, verbose_name=_("created by"))
+    modified_by = models.ForeignKey(
+        settings.AUTH_USER_MODEL, blank=True, null=True,
+        related_name="+", on_delete=models.SET_NULL, verbose_name=_("modified by"))
+    created_on = models.DateTimeField(auto_now_add=True, editable=False, verbose_name=_("created on"))
+    modified_on = models.DateTimeField(auto_now=True, editable=False, verbose_name=_("modified on"))
+
+    # Basic availability
+    shops = models.ManyToManyField(
+        "shuup.Shop", blank=True, db_index=True, verbose_name=_("shops"))
+    active = models.BooleanField(
+        default=False, verbose_name=_("active"),
+        help_text=_("Check this if the discount is currently active. Please also set a start and end date."))
+    start_datetime = models.DateTimeField(
+        null=True, blank=True, verbose_name=_("start date and time"),
+        help_text=_(
+            "The date and time the discount starts. This is only applicable if the discount is marked as active."
+        ))
+    end_datetime = models.DateTimeField(
+        null=True, blank=True, verbose_name=_("end date and time"),
+        help_text=_(
+            "The date and time the discount ends. This is only applicable if the discount is marked as active."
+        ))
+
+    # Conditions
+    product = models.ForeignKey(
+        "shuup.Product", related_name="product_discounts", blank=True, null=True,
+        db_index=True, on_delete=models.CASCADE, verbose_name=_("product"))
+    exclude_selected_category = models.BooleanField(default=False, verbose_name=_("exclude selected category"))
+    category = models.ForeignKey(
+        "shuup.Category", related_name="category_discounts", blank=True, null=True,
+        db_index=True, on_delete=models.CASCADE, verbose_name=_("category"))
+    contact = models.ForeignKey(
```

Comment:
could this be a M2M?

Reply:
I guess the idea of all this is to prevent joining multiple tables but to get results "instantly" from this one table.